rs/p2p/artifact_downloader/src/fetch_artifact/download.rs

```diff
@@ -0,0 +1,293 @@
+    fn clone(&self) -> Self {
+        Self {
+            log: self.log.clone(),
+            transport: self.transport.clone(),
+            priority_fn: self.priority_fn.clone(),
+            metrics: self.metrics.clone(),
+            jh: self.jh.clone(),
+        }
+    }
+}
+
+impl<Artifact: PbArtifact> ArtifactAssembler<Artifact, Artifact> for FetchArtifact<Artifact> {
+    fn disassemble_message(&self, msg: Artifact) -> Artifact {
+        msg
+    }
+    async fn assemble_message<P: Peers + Send + 'static>(
+        &self,
+        id: <Artifact as IdentifiableArtifact>::Id,
+        attr: <Artifact as IdentifiableArtifact>::Attribute,
+        artifact: Option<(Artifact, NodeId)>,
+        peer_rx: P,
+    ) -> Result<(Artifact, NodeId), Aborted> {
+        Self::download_artifact(
+            self.log.clone(),
+            id,
+            attr,
+            artifact,
+            peer_rx,
+            self.priority_fn.clone(),
+            self.transport.clone(),
+            self.metrics.clone(),
+        )
+        .await
+    }
+}
+
+impl<Artifact: PbArtifact> FetchArtifact<Artifact> {
+    pub fn new<Pool>(
+        log: ReplicaLogger,
+        rt: Handle,
+        pool: Arc<RwLock<Pool>>,
+        pfn_producer: Arc<dyn PriorityFnFactory<Artifact, Pool>>,
+        metrics_registry: MetricsRegistry,
+    ) -> (impl Fn(Arc<dyn Transport>) -> Self, Router)
+    where
+        Pool: ValidatedPoolReader<Artifact> + Send + Sync + 'static,
+    {
+        let pool_clone = pool.clone();
+        (
+            move |transport: Arc<dyn Transport>| {
+                let pfn = {
+                    let p = pool.read().unwrap();
+                    pfn_producer.get_priority_function(&p)
+                };
+                let (pfn_tx, pfn_rx) = watch::channel(pfn);
+                let pool_clone = pool.clone();
+                let pfn_producer_clone = pfn_producer.clone();
+                let log_clone = log.clone();
+                let jh = rt.spawn(async move {
+                    loop {
+                        let pfn = {
+                            let p = pool_clone.read().unwrap();
+                            pfn_producer_clone.get_priority_function(&p)
+                        };
+                        if pfn_tx.send(pfn).is_err() {
+                            break;
+                        }
+
+                        tokio::time::sleep(PRIORITY_FUNCTION_UPDATE_INTERVAL).await
+                    }
+                });
+                Self {
+                    log: log_clone,
+                    transport,
+                    priority_fn: pfn_rx,
+                    metrics: FetchArtifactMetrics::new::<Artifact>(&metrics_registry),
+                    jh: Arc::new(jh),
+                }
+            },
+            build_axum_router(pool_clone),
+        )
+    }
+    /// Waits until advert resolves to fetch. If all peers are removed or priority becomes drop `DownloadStopped` is returned.
```

Comment:
This comment seems stale? 
```suggestion
    /// Waits until advert resolves to fetch. If all peers are removed or priority becomes drop `Err(Aborted)` is returned.
```